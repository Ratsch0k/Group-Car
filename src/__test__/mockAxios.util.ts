jest.mock('axios');
import axios from 'axios';
const mockedAxios = axios as jest.Mocked<typeof axios>;

mockedAxios.head.mockImplementationOnce((path: string) => {
  if (path === '/auth') {
    return Promise.resolve({
      headers: {
        'xsrf-token': 'TEST TOKEN',
      },
    });
  }
});

export default mockedAxios;
