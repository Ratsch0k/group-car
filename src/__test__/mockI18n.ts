jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => undefined),
        languages: ['de', 'en'],
        language: 'en',
      },
    };
  },
}));
