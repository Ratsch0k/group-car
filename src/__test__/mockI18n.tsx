import React from 'react';

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => undefined),
        languages: ['de', 'en'],
        language: 'en',
      },
    };
  },
  Trans: (props: any) => {
    return <div>{props.i18nKey}</div>;
  },
}));
