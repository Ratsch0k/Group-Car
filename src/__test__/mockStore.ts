/* eslint-disable @typescript-eslint/ban-types */
import {RootState} from '../lib/redux/store';
import thunk, {ThunkDispatch} from 'redux-thunk';
import {
  initialState as authState,
} from '../lib/redux/slices/auth/authSlice';
import {
  initialState as groupState,
} from '../lib/redux/slices/group/groupSlice';
import {
  initialState as invitesState,
} from '../lib/redux/slices/invites/invitesSlice';
import configureMockStore, {
  MockStoreCreator,
  MockStoreEnhanced,
} from 'redux-mock-store';
import {AnyAction, Middleware} from '@reduxjs/toolkit';

const defaultState: RootState = {
  router: {
    action: 'POP',
    location: {
      pathname: '/',
      search: '',
      hash: '',
      query: {},
      state: undefined,
    },
  },
  auth: authState,
  group: groupState,
  invites: invitesState,
};

type DispatchExts = ThunkDispatch<RootState, undefined, AnyAction>;


/**
 * Creates a mock store.
 * @param state Initial state. Will be merged with the default initial state.
 * @returns The mocked store
 */
function mockStore(state?: Partial<RootState>):
MockStoreEnhanced<RootState, DispatchExts> {
  const middlewares: Array<Middleware> = [thunk];
  const mockStoreCreator: MockStoreCreator<RootState, DispatchExts> =
  configureMockStore<RootState, DispatchExts>(middlewares);
  const initialState: RootState = {
    ...defaultState,
    ...state,
  };
  const store = mockStoreCreator(initialState);

  return store;
}

export default mockStore;
