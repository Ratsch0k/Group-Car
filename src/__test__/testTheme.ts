import {createMuiTheme} from '@material-ui/core';
import {themeProperties, IThemeOptions} from '../lib/theme';

export default createMuiTheme({
  ...themeProperties,
  props: {
    MuiButtonBase: {
      disableTouchRipple: true,
      disableRipple: true,
    },
  },
} as IThemeOptions);
