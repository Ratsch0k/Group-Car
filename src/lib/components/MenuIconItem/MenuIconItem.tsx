import React from 'react';
import {MenuItemProps, MenuItem, Theme} from '@material-ui/core';
import {MenuItemToSubMenu} from 'lib';
import {makeStyles, createStyles} from '@material-ui/styles';

/**
 * Props for the menu icon item element.
 */
export interface MenuIconItemProps extends MenuItemProps {
  /**
   * If this menu item should look like it opens a
   * sub menu. An arrow will be displayed on the right side.
   */
  opensSubMenu?: boolean;

  /**
   * The icon which should be displayed on the left.
   */
  icon?: JSX.Element;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    text: {
      paddingLeft: theme.spacing(1),
    },
  }),
);

/**
 * Combines `MenuItem` and `MenuItemToSubMenu` and
 * provides option to append the menu item with
 * an icon.
 * @param props Props.
 */
export const MenuIconItem: React.FC<MenuIconItemProps> = (props) => {
  const {opensSubMenu, icon, children, ...rest} = props;
  const classes = useStyles();

  let Item;
  if (opensSubMenu) {
    Item = MenuItemToSubMenu;
  } else {
    Item = MenuItem;
  }

  return (
    <Item {...rest as unknown}>
      {icon}
      <span className={icon ? classes.text : undefined}>
        {children}
      </span>
    </Item>
  );
};

export default MenuIconItem;
