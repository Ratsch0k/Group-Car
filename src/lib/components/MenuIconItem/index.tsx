export * from './MenuIconItem';

export {default} from './MenuIconItem';
