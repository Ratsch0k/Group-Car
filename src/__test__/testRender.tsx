import React from 'react';
import {render, RenderResult} from '@testing-library/react';
import {ConnectedRouter} from 'connected-react-router';
import {Provider} from 'react-redux';
import history from '../lib/redux/history';
import {StylesProvider, ThemeProvider} from '@material-ui/styles';
import testTheme from './testTheme';
import configureMockStore, {
  MockStoreCreator,
  MockStoreEnhanced,
} from 'redux-mock-store';
import thunk from 'redux-thunk';
import {RootState} from '../lib/redux/store';
import {
  initialState as authState,
} from '../lib/redux/slices/auth/authSlice';
import {
  initialState as groupState,
} from '../lib/redux/slices/group/groupSlice';
import {
  initialState as invitesState,
} from '../lib/redux/slices/invites/invitesSlice';


export interface TestRenderResult<T> {
  mockStore: MockStoreCreator<T>;
  store: MockStoreEnhanced<unknown>;
}

const defaultState: RootState = {
  router: {
    action: 'POP',
    location: {
      pathname: '/',
      search: '',
      hash: '',
      query: {},
      state: undefined,
    },
  },
  auth: authState,
  group: groupState,
  invites: invitesState,
};

/**
 * Testing generateClassName function to
 * avoid weird fails due to random class names
 * @param rule Rule
 * @param stylesheet Stylesheet
 * @returns The class name
 */
const generateClassName = (rule, stylesheet) => {
  return `${stylesheet.options.classNamePrefix}-${rule.key}`;
};


/**
 * Render function for testing purposes.
 * Provides the necessary base components and
 * contexts.
 * @param children The children to render
 * @returns The result of the render function.
 */
function testRender<T>(state: T, children: React.ReactNode):
RenderResult & TestRenderResult<T> {
  const middleware = [thunk];
  const mockStore = configureMockStore<T>(middleware);
  const initialState = {
    ...defaultState,
    ...state,
  };
  const store = mockStore(initialState);

  const renderResult = render(
    <Provider store={store}>
      <ConnectedRouter history={history}>
        <ThemeProvider theme={testTheme}>
          <StylesProvider generateClassName={generateClassName}>
            {children}
          </StylesProvider>
        </ThemeProvider>
      </ConnectedRouter>
    </Provider>,
  );

  return {
    ...renderResult,
    mockStore,
    store,
  };
}

export default testRender;
