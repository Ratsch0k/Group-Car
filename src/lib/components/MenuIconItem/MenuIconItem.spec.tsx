import testRender from '../../../__test__/testRender';
import '../../../__test__/mockAxios';
import MenuIconItem from "./MenuIconItem";
import React from "react";
import AddIcon from '@material-ui/icons/Add'

it('renders without crashing', () => {
  testRender(
    {},
      <MenuIconItem>
      TEST
      </MenuIconItem>
  );
});

it('matches MenuItem if no property provided', () => {
  const {container} = testRender(
    {},
      <MenuIconItem>
        TEST
      </MenuIconItem>
  );
  
  expect(container).toMatchSnapshot();
});

it('displays icon on the left if icon is provided', () => {
  const {container} = testRender(
    {},
      <MenuIconItem icon={<AddIcon />}>
        TEST
      </MenuIconItem>
  );
  
  expect(container).toMatchSnapshot();
});

it('shows arrow to the right if prop opensSubMenu is true', () => {
  const {container} = testRender(
    {},
      <MenuIconItem opensSubMenu>
        TEST
      </MenuIconItem>
  );
  
  expect(container).toMatchSnapshot();
});

it('shows arrow to the right if prop opensSubMenu is true and icon on the left if icon is provided', () => {
  const {container} = testRender(
    {},
      <MenuIconItem opensSubMenu icon={<AddIcon />}>
        TEST
      </MenuIconItem>
  );
  
  expect(container).toMatchSnapshot();
});