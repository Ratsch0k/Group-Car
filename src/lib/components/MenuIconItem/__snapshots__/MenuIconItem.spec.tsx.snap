// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays icon on the left if icon is provided 1`] = `
<div>
  <li
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
    role="menuitem"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
      />
    </svg>
    <span
      class="makeStyles-text-59"
    >
      TEST
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </li>
</div>
`;

exports[`matches MenuItem if no property provided 1`] = `
<div>
  <li
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
    role="menuitem"
    tabindex="-1"
  >
    <span>
      TEST
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </li>
</div>
`;

exports[`shows arrow to the right if prop opensSubMenu is true 1`] = `
<div>
  <li
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
    role="menuitem"
    tabindex="-1"
  >
    <span>
      TEST
    </span>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root makeStyles-arrowIcon-98"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M5.88 4.12L13.76 12l-7.88 7.88L8 22l10-10L8 2z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root"
    />
  </li>
</div>
`;

exports[`shows arrow to the right if prop opensSubMenu is true and icon on the left if icon is provided 1`] = `
<div>
  <li
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
    role="menuitem"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
      />
    </svg>
    <span
      class="makeStyles-text-136"
    >
      TEST
    </span>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root makeStyles-arrowIcon-137"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M5.88 4.12L13.76 12l-7.88 7.88L8 22l10-10L8 2z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root"
    />
  </li>
</div>
`;
